src/main/scala/eu/proteus/solma/lasso/LassoWorkerLogic.scala

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright (C) 2017 The Proteus Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package eu.proteus.solma.lasso
+
+import breeze.linalg.{DenseVector => DenseBreezeVector, Vector => BreezeVector}
+import org.apache.flink.ml.math.Breeze._
+import eu.proteus.solma.events.StreamEventWithPos
+import eu.proteus.solma.lasso.Lasso.{LassoModel, LassoParam, OptionLabeledVector}
+import eu.proteus.solma.lasso.LassoStreamEvent.LassoStreamEvent
+import eu.proteus.solma.lasso.algorithm.{FlatnessMappingAlgorithm, LassoAlgorithm, LassoBasicAlgorithm}
+import hu.sztaki.ilab.ps.{ParameterServerClient, WorkerLogic}
+
+import scala.collection.mutable
+
+class LassoWorkerLogic (modelBuilder: ModelBuilder[LassoParam, LassoModel],
+                        lassoMethod: LassoAlgorithm[OptionLabeledVector, LassoParam, ((Long, Double), Double),
+                          LassoModel]) extends WorkerLogic[LassoStreamEvent, LassoParam, ((Long, Double), Double)]
+{
+  val unpredictedVecs = new mutable.Queue[LassoStreamEvent]()
+  val unlabeledVecs = new mutable.HashMap[Long, mutable.Queue[StreamEventWithPos[(Long, Double)]]]()
+  val labeledVecs = new mutable.Queue[OptionLabeledVector]
+
+  override def onRecv(data: LassoStreamEvent,
+                      ps: ParameterServerClient[LassoParam, ((Long, Double), Double)]): Unit = {
+    data match {
+      case Left(v) =>
+        if (!unlabeledVecs.keys.exists(x => x == v.pos._1)) {
+          unlabeledVecs(v.pos._1) = new mutable.Queue[StreamEventWithPos[(Long, Double)]]()
+        }
+        unlabeledVecs(v.pos._1).enqueue(v)
+        unpredictedVecs.enqueue(data)
+        ps.pull(0)
+      case Right(v) =>
+        if (unlabeledVecs.keys.exists(x => x == v.label)) {
```

Comment:
The logic seems correct and I am quite sure it should work on the real scenario, i.e., when measurements for a given coil arrive earlier than the true flatness. However, running your test with a debugger showed that the true flatness arrives somehow earlier than the measurements. This is due to the overall asynchronous processing. In particular, what fails in your code is line 48. Basically, the if fails because the true flatness arrive first. The solution would be to either cache the prediction (although it seems a bit unlikely to happen in production) or to delay the flatness stream in the UT. What do you think? Which option do you think it is easier to implement?

Reply:
I have added changes to build a list with all events (features and true flatness) and where features are before true flatness in the list. If this list of events are the source of a DataStream and parallelism is set to one, the processing order is guaranteed?

Reply:
The order-policy within a a stream with parallelism of 1 is FIFO. If a stream has parallelism > 1, then each sub-stream/channel is FIFO, but the order is not preserver among sub-streams. When you connect two DataStream-s with an arbitrary parallelism, then the order is preserved only within every channel, i.e., FIFO. This means we cannot control what arrives first in a coMap/coFlatMap. This also means that we cannot control the arrival order in the onRecv of the parameter server.

If you stream a list like you said then yes, the order should be preserved as long as the parallelism is kept constant. If there is a change of parallelism or a keyBy, then the order is altered. 

Please, let me know if  you have any further questions and how I can help you in finalising this PR :)

Reply:
I have added more changes and I think PR is ready